Engine/src/Engine/Serializer.cpp

```diff
@@ -0,0 +1,181 @@
+	{
+		if (!out.initializeGL())
+		{
+			GE_CORE_FATAL("Failed to initialize opengl");
+			return false;
+		}
+
+		auto data = ResourceManager::getInstance()->getJsonData(sceneFile);
+		if (!(data.find("scene") != data.end()))
+		{
+			GE_CORE_FATAL("Unable to deserialize scene {0}", sceneFile);
+			GE_CORE_FATAL("No scene found.");
+			return false;
+		}
+
+		if (!(data["scene"].find("name") != data["scene"].end()))
+		{
+			GE_CORE_FATAL("Unable to deserialize scene {0}", sceneFile);
+			GE_CORE_FATAL("Scene does not contain the \"name\" attribute.");
+			return false;
+		}
+
+		out.m_name = data["scene"]["name"];
+
+		//Loop through all entities and deserialize each
+		for (const auto& item : data["scene"]["entities"])
+		{
+			if (!(item.find("tag") != item.end()))
+			{
+				GE_CORE_FATAL("Unable to deserialize scene {0}", sceneFile);
+				GE_CORE_FATAL("An entity does not contain the \"tag\" attribute");
+				return false;
+			}
+
+			Entity entity = out.createEntity(item["tag"]);
+			if (!tryDeserializeEntity(entity, item, out))
+			{
+				GE_CORE_FATAL("Unable to deserialize scene {0}", sceneFile);
+				GE_CORE_FATAL("The entity {0} has failed to serialize", item["tag"]);
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	void Serializer::serializeScene(const Scene& scene, const std::string& sceneFile)
+	{
+		//TODO: Implement serialization once we have more GUI stuff
+	}
+
+	bool Serializer::tryDeserializeEntity(Entity& out, const nlohmann::json& entity, Scene& scene)
+	{
+		//Loop through all components and deserialize each
+		for (const auto& component : entity["components"])
+		{
+			if (!(component.find("name") != component.end()))
+			{
+				GE_CORE_FATAL("Unable to deserialize entity {0}", entity["tag"]);
+				GE_CORE_FATAL("A component does not contain the \"name\" attribute");
+				return false;
+			}
+
+			auto name = component["name"].get<std::string>();
+
+			//check each component type individually
+			if (name == "CameraComponent")
+			{
+				auto fov = component["fov"].get<float>();
+				auto projection = component["projection"].get<glm::mat4>();
+				auto viewport = component["viewport"].get<glm::vec2>();
+				auto farZ = component["farZ"].get<float>();
+				auto nearZ = component["nearZ"].get<float>();
+
+				out.addComponent<CameraComponent>(fov, projection, viewport, farZ, nearZ);
+			}
+
+			else if (name == "TransformComponent")
+			{
+				auto position = component["position"].get<glm::vec3>();
+				auto scale = component["scale"].get<glm::vec3>();
+				auto rotation = component["rotation"].get<glm::vec3>();
+
+				out.addComponent<TransformComponent>(position, scale, rotation);
+			}
+		
+			else if (name == "ColorComponent")
+			{
+				out.addComponent<ColorComponent>(component["color"].get<glm::vec4>());
+			}
+		
+			else if (name == "TextureComponent")
+			{
+				std::string texture = component["texName"];
+				auto image = ResourceManager::getInstance()->getTexture(texture);
```

Comment:
Needs error checking if texture file not found. If no file is found, getTexture will return an empty ImageData object.

Reply:
This is already the case 
![image](https://user-images.githubusercontent.com/9102977/198858103-e15f7a91-dc54-4506-a7d0-da39710b5aac.png)

Reply:
I am aware that it logs the error, but in my opinion there should be special handling for a texture not found during the deserialization process. Perhaps we could use default error texture like many other engines do. Otherwise, I believe that failing to find a texture should fail the deserialize function. If loading the texture fails but the empty image data is still passed to the component, the call to bind texture will bind texture 0, which is an unbind call. I feel like this may open the door to a future bug that could be difficult to trace. Having said that, it is not an immediately pressing concern.

Reply:
Yeah I agree with that. It would be great to have more error handling, but I believe that's a rendering issue and possibly out of this PR's scope